design.md

```diff
@@ -0,0 +1,254 @@
+
+TLS 1.3 will be used to ensure a secure cipher suite is utilized. As of Go 1.17, the cipher suites are ordered by `crypto/tls` based on range of factors. https://go.dev/blog/tls-cipher-suites
+
+A set of certificates and secrets will exist in the `certs/` directory. All certificates will be signed by the included CA.
+
+## Authorization
+
+Once a client establishes a connection over TLS, the client certificate's `Common Name` will be used to determine which user has connected. A user will only be authorized to interact with jobs they started. All other jobs will be inaccessible to the user's client.
+
+## Critical Libraries
+
+### cgroups
+
+Package cgroups will provide mechanisms to setup, cleanup, and interact with Linux cgroups.
+
+The `Service` type will be responsible for the following:
+- mounting/unmounting cgroup2
+- directory `/cgroup2/jobworker` setup/cleanup
+- creating/removing a cgroup
+- updating cgroup controllers
+- placing a process into a cgroup
+
+#### Types
+
+```
+type Service struct {
+  ID uuid.UUID
+}
+
+func (s Service) Cleanup() error
+func (s Service) CreateCgroup(options ...CgroupOption) (*Cgroup, error)
+func (s Service) RemoveCgroup(id uuid.UUID) error
+func (s Service) PlaceInCgroup(cgroup Cgroup, pid int) error
+
+type Cgroup struct {
+  ID             uuid.UUID
+  Memory         string
+  Cpus           float32
+  DeviceWriteBps string
+  DeviceReadBps  string
+}
+
+type CgroupOption func(*Cgroup)
+
+func WithMemory(limit string) CgroupOption
+func WithCPUs(cpus float32) CgroupOption
+func WithDeviceWriteBps(deviceLimit string) CgroupOption
+func WithDeviceReadBps(deviceLimit string) CgroupOption
+...
+```
+
+---
+
+### job
+
+Package job will provide a `Service` type for managing `Job` instances.
+
+The `Service` type will be responsible for the following.
+- Setup and cleanup of `var/log/jobworker` logging directory.
+- Maintaining a concurrent-safe jobs map with `Job.ID` as the key and `Job` as the value.
+- Provide methods to start, stop, and fetch managed `Job` instances
+
+The `Job` type will be responsible for the following.
+- Provide unexported mechanisms to start and stop arbitrary commands.
+- Provide exported mechanisms to stream output to multiple clients concurrently.
+- Ensure thread safe access to attributes where necessary.
+- Ensure status reflects command state.
+
+##### Start, Stop, and Status
+
+The starting, stopping and monitoring of a `Job` instances command will be done mostly by utilizing the `os/exec` package.
+
+In order to track command status, start (`os/exec.Cmd.Start()`) will be called and a goroutine will be launched to wait (`os/exec.Cmd.Wait()`) for command completion and record exit status.
+
+##### Streaming Output
+
+Streaming output will involve reading a log file from `/var/log/jobworker` in chunks and writing it to the client until EOF is reached. If the client uses the `--tail` flag, the output will be streamed until the client ends the stream.
```

Comment:
You can always `--tail` to simplify. Also can you describe how you will implement streaming?

Reply:
I've removed the `--tail` flag and I've added details on how streaming would be implemented.

Reply:
How will you handle reading from the file when the a read returns an EOF error, yet the job is still running and may produce more output in the future?

Reply:
I've added this detail to the doc under **Stream Output - Implementation**.